File: 递归实现排列型枚举.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    int n;
    cin >> n;
    vector<int> st(n + 1),path;

    function<void()> dfs = [&]() {
        if (path.size() == n) {
            for (int pi : path) {
                cout << pi << " ";
            }
            cout << endl;
            return;
        }
        for (int i = 1; i <= n; ++i) {
            if (not st[i]) {
                st[i] = 1;
                path.emplace_back(i);
                dfs();
                path.pop_back();
                st[i] = 0;
            }
        }
    };

    dfs();
    return 0;
}

File: 递归实现排列型枚举.py
n = int(input())
path = []
st = [0 for _ in range(n + 1)]

def dfs():
    if len(path) == n:
        print(*path)
        return
    for i in range(1, n + 1):
        if not st[i]:
            st[i] = 1
            path.append(i)
            dfs()
            path.pop()
            st[i] = 0

dfs()
